### PR TITLE
Transaction validation and filtering

### PR DESCRIPTION
**Pull request by bowd**
### Overview

We need a solution to decode the meta transaction payload in two scenarios:
- when requesting the subsidy attestation we want to ensure that the two txs passed in are specifically an cUSD.approve and Attestations.request
- when passing in an arbitrary meta-tx it is validated agains the allowed transactions.

### CODE REVIEW

File: apps/onboarding/src/wallet/wallet.service.ts [hunk 1]
```diff
@@ -123,4 +170,29 @@ export class WalletService {
   private isValidImplementation(implementationAddress: string): boolean {
     return implementationAddress in this.cfg.mtwImplementations
   }
+
+  private decodeMetaTransaction(tx: RawTransaction): Result<RawTransaction, MetaTxValidationError> {
+    // const wallet = await this.contractKit.contracts.getMetaTransactionWallet(tx.destination)
```
Comment (codyborn) [inline apps/onboarding/src/wallet/wallet.service.ts:175]: nit: remove leftover

File: apps/onboarding/src/wallet/wallet.service.ts [hunk 2]
```diff
@@ -123,4 +170,29 @@ export class WalletService {
...
+
+    if (decodedData.method === null) {
+      return Err(new InputDecodeError())
+    }
+
+    if (decodedData.method !== 'executeMetaTransaction') {
+      return Err(new InvalidRootMethod(decodedData.method))
+    }
+
+    const inputs: [string, any, Buffer] = decodedData.inputs
```
Comment (codyborn) [inline apps/onboarding/src/wallet/wallet.service.ts:191]: We should verify the length of inputs here

File: apps/onboarding/src/wallet/wallet.service.ts [hunk 3]
```diff
@@ -42,17 +41,65 @@ export class WalletService {
...
+    const txWithMatchingDestination = allowedTransactions.filter(
+      (tx) =>
+        normalizeAddress(tx.destination) === normDestination
+    )
+
+    if (txWithMatchingDestination.length === 0) {
+      return Err(new InvalidDestination(metaTx.destination))
+    }
+
+    const metaTxMethodId = trimLeading0x(metaTx.data).slice(0, 8)
```
Comment (codyborn) [inline apps/onboarding/src/wallet/wallet.service.ts:65]: Can you add a comment to explain where this `.slice(0, 8)` comes from?

Comment (codyborn) [reply apps/onboarding/src/wallet/wallet.service.ts:65]: optionally can wrap it in a util method to make it more self-documenting

File: apps/onboarding/src/wallet/wallet.service.ts [hunk 4]
```diff
@@ -42,17 +41,65 @@ export class WalletService {
...
+    const matchingTx = txWithMatchingDestination.find(
+      (tx) => {
+        const methodId = trimLeading0x(tx.methodId)
+        return metaTxMethodId === methodId
```
Comment (codyborn) [inline apps/onboarding/src/wallet/wallet.service.ts:69]: Any normalization required before comparison? (ex. toLowerCase)

File: apps/onboarding/src/subsidy/subsidy.service.ts [hunk 5]
```diff
@@ -20,23 +22,44 @@ export class SubsidyService {
...
+
+    if (approveValid.ok === false) {
+      return approveValid
     }
 
-    if (normalizeAddress(input.transactions.request.destination) !== normalizeAddress(input.walletAddress)) {
-      return Err(new InvalidMetaTransaction("subsidyRequest:request"))
+    const attestations = await this.contractKit.contracts.getAttestations()
+    const requestValid = await this.walletService.isAllowedMetaTransaction(
+      input.transactions.approve,
```
Comment (codyborn) [inline apps/onboarding/src/subsidy/subsidy.service.ts:52]:
```suggestion
      input.transactions.request,
```

File: apps/onboarding/src/wallet/wallet.service.spec.ts [hunk 6]
```diff
@@ -0,0 +1,503 @@
...
+          provide: appConfig.KEY,
+          useValue: cfg
+        },
+      ]
+    }).compile()
+  }
+
+  describe('#isAllowedMetaTransaction', () => {
+    describe('with a random transaction', () => {
+      it('returns an DecodeError', async () => {
```
Comment (codyborn) [inline apps/onboarding/src/wallet/wallet.service.spec.ts:58]:
```suggestion
      it('returns a DecodeError', async () => {
```

**Review by codyborn — APPROVED** [review]
Nice work!

File: apps/onboarding/src/wallet/wallet.service.ts (hunk 4 above)
Comment (bowd) [reply apps/onboarding/src/wallet/wallet.service.ts:69]: Currently no, but I guess it's worth adding just to be sure that nothing changes somewhere down the line at any point and this breaks.

**Pull request merged**